File: typescript/base-hmi-types/src/projects/HmiProjectBase.ts
import { HmiAlarmList } from "../alarms/HmiAlarm.js";
import { HmiConnectionList } from "../connections/HmiConnectionList.js";
import { HmiCycle } from "../cycles/HmiCycle.js";
import { HmiImage } from "../images/HmiImage.js";
import { HmiFaceplateType } from "../screens/base/HmiFaceplateType.js";
import { HmiScreenBase } from "../screens/base/HmiScreenBase.js";
import { HmiScript } from "../scripts/HmiScript.js";
import { HmiTagTable } from "../tags/HmiTagTable.js";
import { HmiGraphicList } from "../text-graphic-lists/HmiGraphicList.js";
import { HmiTextList } from "../text-graphic-lists/HmiTextList.js";
import { IHmiProject, IHmiProjectFolder } from "./IHmiProject.js";
import { HmiProjectFolderType } from "./HmiProjectFolderType.js";
import { IHmiProjectItemDescriptor } from "./HmiProjectItemDescriptor.js";
import { HmiProjectInfo } from "./HmiProjectInfo.js";
import { HmiProjectItemKind } from "./HmiProjectItemKind.js";
import { HmiScreenDescriptor } from "./HmiScreenDescriptor.js";

export abstract class HmiProjectBase implements IHmiProject {
  readonly info = new HmiProjectInfo();

  get id(): string | undefined {
    return undefined;
  }

  get name(): string {
    return this.info.projectName?.trim() ? this.info.projectName : "Project";
  }

  get path(): string {
    return this.name;
  }

  get folderType(): HmiProjectFolderType {
    return HmiProjectFolderType.Unknown;
  }

  get folders(): readonly IHmiProjectFolder[] {
    return [];
  }

  get items(): readonly IHmiProjectItemDescriptor[] {
    return [];
  }

  get screens(): IHmiProjectFolder | undefined {
    return this.getFolder(HmiProjectFolderType.Screens);
  }

  get faceplates(): IHmiProjectFolder | undefined {
    return this.getFolder(HmiProjectFolderType.Faceplates);
  }

  get tags(): IHmiProjectFolder | undefined {
    return this.getFolder(HmiProjectFolderType.Tags);
  }

  get scripts(): IHmiProjectFolder | undefined {
    return this.getFolder(HmiProjectFolderType.Scripts);
  }

  get textLists(): IHmiProjectFolder | undefined {
    return this.getFolder(HmiProjectFolderType.TextLists);
  }

  get graphicLists(): IHmiProjectFolder | undefined {
    return this.getFolder(HmiProjectFolderType.GraphicLists);
  }

  get images(): IHmiProjectFolder | undefined {
    return this.getFolder(HmiProjectFolderType.Images);
  }

  get cycles(): IHmiProjectFolder | undefined {
    return this.getFolder(HmiProjectFolderType.Cycles);
  }

  get alarms(): IHmiProjectFolder | undefined {
    return this.getFolder(HmiProjectFolderType.Alarms);
  }

  get connections(): IHmiProjectFolder | undefined {
    return this.getFolder(HmiProjectFolderType.Connections);
  }

  getFolders(_signal?: AbortSignal): Promise<readonly IHmiProjectFolder[]> {
    return Promise.resolve([]);
  }

  getItems(_signal?: AbortSignal): Promise<readonly IHmiProjectItemDescriptor[]> {
    return Promise.resolve([]);
  }

  getScreen(_id: string, _signal?: AbortSignal): Promise<HmiScreenBase | undefined> {
    return Promise.resolve(undefined);
  }

  getFaceplate(_id: string, _signal?: AbortSignal): Promise<HmiFaceplateType | undefined> {
    return Promise.resolve(undefined);
  }

  getFaceplateByNameAndVersion(
    _name: string,
    _version: string,
    _signal?: AbortSignal,
  ): Promise<HmiFaceplateType | undefined> {
    return Promise.resolve(undefined);
  }

  getTagTable(_id: string, _signal?: AbortSignal): Promise<HmiTagTable | undefined> {
    return Promise.resolve(undefined);
  }

  getScript(_id: string, _signal?: AbortSignal): Promise<HmiScript | undefined> {
    return Promise.resolve(undefined);
  }

  getTextList(_id: string, _signal?: AbortSignal): Promise<HmiTextList | undefined> {
    return Promise.resolve(undefined);
  }

  getGraphicList(_id: string, _signal?: AbortSignal): Promise<HmiGraphicList | undefined> {
    return Promise.resolve(undefined);
  }

  getImage(_id: string, _signal?: AbortSignal): Promise<HmiImage | undefined> {
    return Promise.resolve(undefined);
  }

  getCycle(_id: string, _signal?: AbortSignal): Promise<HmiCycle | undefined> {
    return Promise.resolve(undefined);
  }

  getAlarmList(_id: string, _signal?: AbortSignal): Promise<HmiAlarmList | undefined> {
    return Promise.resolve(undefined);
  }

  getConnectionList(_id: string, _signal?: AbortSignal): Promise<HmiConnectionList | undefined> {
    return Promise.resolve(undefined);
  }

  async getScreens(signal?: AbortSignal): Promise<readonly HmiScreenDescriptor[]> {
    const result: HmiScreenDescriptor[] = [];
    const screens = this.screens;
    if (screens === undefined) {
      return result;
    }

    await collectScreens(screens, result, signal);
    return result;
  }

  protected getFolder(folderType: HmiProjectFolderType): IHmiProjectFolder | undefined {
    return this.folders.find((folder) => folder.folderType === folderType);
  }
}

async function collectScreens(
  folder: IHmiProjectFolder,
  result: HmiScreenDescriptor[],
  signal?: AbortSignal,
): Promise<void> {
  for (const item of await folder.getItems(signal)) {
    if (item.kind === HmiProjectItemKind.Screen) {
      const descriptor = new HmiScreenDescriptor();
      descriptor.id = item.id;
      descriptor.name = item.name;
      descriptor.displayName = item.name;
      descriptor.path = item.path;
      descriptor.sourceType = item.sourceType;
      result.push(descriptor);
    }
  }

  for (const child of await folder.getFolders(signal)) {
    await collectScreens(child, result, signal);
  }
}

File: src/BaseHmiTypes/Projects/HmiProjectBase.cs
using BaseHmiTypes.Alarms;
using BaseHmiTypes.Connections;
using BaseHmiTypes.Cycles;
using BaseHmiTypes.Images;
using BaseHmiTypes.Screens;
using BaseHmiTypes.Scripts;
using BaseHmiTypes.Tags;
using BaseHmiTypes.TextGraphicLists;

namespace BaseHmiTypes.Projects;

public abstract class HmiProjectBase : IHmiProject
{
    public HmiProjectInfo Info { get; } = new();

    public virtual string? Id => null;

    public virtual string Name => string.IsNullOrWhiteSpace(Info.ProjectName) ? "Project" : Info.ProjectName!;

    public virtual string Path => Name;

    public virtual HmiProjectFolderType FolderType => HmiProjectFolderType.Unknown;

    public IReadOnlyList<IHmiProjectFolder> Folders => GetFoldersAsync().AsTask().GetAwaiter().GetResult();

    public IReadOnlyList<IHmiProjectItemDescriptor> Items => GetItemsAsync().AsTask().GetAwaiter().GetResult();

    public virtual IHmiProjectFolder? Screens => GetFolder(HmiProjectFolderType.Screens);

    public virtual IHmiProjectFolder? Tags => GetFolder(HmiProjectFolderType.Tags);

    public virtual IHmiProjectFolder? Scripts => GetFolder(HmiProjectFolderType.Scripts);

    public virtual IHmiProjectFolder? TextLists => GetFolder(HmiProjectFolderType.TextLists);

    public virtual IHmiProjectFolder? GraphicLists => GetFolder(HmiProjectFolderType.GraphicLists);

    public virtual IHmiProjectFolder? Images => GetFolder(HmiProjectFolderType.Images);

    public virtual IHmiProjectFolder? Cycles => GetFolder(HmiProjectFolderType.Cycles);

    public virtual IHmiProjectFolder? Alarms => GetFolder(HmiProjectFolderType.Alarms);

    public virtual IHmiProjectFolder? Connections => GetFolder(HmiProjectFolderType.Connections);

    public virtual ValueTask<IReadOnlyList<IHmiProjectFolder>> GetFoldersAsync(CancellationToken cancellationToken = default)
    {
        return new ValueTask<IReadOnlyList<IHmiProjectFolder>>(Array.Empty<IHmiProjectFolder>());
    }

    public virtual ValueTask<IReadOnlyList<IHmiProjectItemDescriptor>> GetItemsAsync(CancellationToken cancellationToken = default)
    {
        return new ValueTask<IReadOnlyList<IHmiProjectItemDescriptor>>(Array.Empty<IHmiProjectItemDescriptor>());
    }

    public virtual ValueTask<HmiScreen?> GetScreenAsync(string id, CancellationToken cancellationToken = default)
    {
        return new ValueTask<HmiScreen?>((HmiScreen?)null);
    }

    public virtual ValueTask<HmiTagTable?> GetTagTableAsync(string id, CancellationToken cancellationToken = default)
    {
        return new ValueTask<HmiTagTable?>((HmiTagTable?)null);
    }

    public virtual ValueTask<HmiScript?> GetScriptAsync(string id, CancellationToken cancellationToken = default)
    {
        return new ValueTask<HmiScript?>((HmiScript?)null);
    }

    public virtual ValueTask<HmiTextList?> GetTextListAsync(string id, CancellationToken cancellationToken = default)
    {
        return new ValueTask<HmiTextList?>((HmiTextList?)null);
    }

    public virtual ValueTask<HmiGraphicList?> GetGraphicListAsync(string id, CancellationToken cancellationToken = default)
    {
        return new ValueTask<HmiGraphicList?>((HmiGraphicList?)null);
    }

    public virtual ValueTask<HmiImage?> GetImageAsync(string id, CancellationToken cancellationToken = default)
    {
        return new ValueTask<HmiImage?>((HmiImage?)null);
    }

    public virtual ValueTask<HmiCycle?> GetCycleAsync(string id, CancellationToken cancellationToken = default)
    {
        return new ValueTask<HmiCycle?>((HmiCycle?)null);
    }

    public virtual ValueTask<HmiAlarmList?> GetAlarmListAsync(string id, CancellationToken cancellationToken = default)
    {
        return new ValueTask<HmiAlarmList?>((HmiAlarmList?)null);
    }

    public virtual ValueTask<HmiConnectionList?> GetConnectionListAsync(string id, CancellationToken cancellationToken = default)
    {
        return new ValueTask<HmiConnectionList?>((HmiConnectionList?)null);
    }

    public virtual async ValueTask<IReadOnlyList<HmiScreenDescriptor>> GetScreensAsync(CancellationToken cancellationToken = default)
    {
        var result = new List<HmiScreenDescriptor>();
        var screens = Screens;
        if (screens == null)
            return result;

        await CollectScreensAsync(screens, result, cancellationToken).ConfigureAwait(false);
        return result;
    }

    protected IHmiProjectFolder? GetFolder(HmiProjectFolderType folderType)
    {
        return Folders.FirstOrDefault(folder => folder.FolderType == folderType);
    }

    private static async ValueTask CollectScreensAsync(
        IHmiProjectFolder folder,
        IList<HmiScreenDescriptor> result,
        CancellationToken cancellationToken)
    {
        foreach (var item in await folder.GetItemsAsync(cancellationToken).ConfigureAwait(false))
        {
            if (item.Kind == HmiProjectItemKind.Screen)
            {
                result.Add(new HmiScreenDescriptor
                {
                    Id = item.Id,
                    Name = item.Name,
                    DisplayName = item.Name,
                    Path = item.Path,
                    SourceType = item.SourceType
                });
            }
        }

        foreach (var child in await folder.GetFoldersAsync(cancellationToken).ConfigureAwait(false))
            await CollectScreensAsync(child, result, cancellationToken).ConfigureAwait(false);
    }
}
